azure-kusto-data/azure/kusto/data/aio/streaming_response.py

```diff
@@ -0,0 +1,248 @@
+
+    async def read_token_of_type(self, *token_types: JsonTokenType) -> JsonToken:
+        token = await self.read_next_token_or_throw()
+        if token.token_type not in token_types:
+            raise KustoTokenParsingError(f"Expected one the following types: '{','.join(t.name for t in token_types)}' , got type {token.token_type}")
+        return token
+
+    async def read_start_object(self) -> JsonToken:
+        return await self.read_token_of_type(JsonTokenType.START_MAP)
+
+    async def read_start_array(self) -> JsonToken:
+        return await self.read_token_of_type(JsonTokenType.START_ARRAY)
+
+    async def read_string(self) -> str:
+        return (await self.read_token_of_type(JsonTokenType.STRING)).token_value
+
+    async def read_boolean(self) -> bool:
+        return (await self.read_token_of_type(JsonTokenType.BOOLEAN)).token_value
+
+    async def read_number(self) -> float:
+        return (await self.read_token_of_type(JsonTokenType.NUMBER)).token_value
+
+    async def skip_children(self, prev_token: JsonToken):
+        if prev_token.token_type == JsonTokenType.MAP_KEY:
+            child_token = await self.read_next_token_or_throw()
+            await self.skip_children(child_token)
+        elif prev_token.token_type in JsonTokenType.start_tokens():
+            async for potential_end_token in self:
+                if potential_end_token.token_path == prev_token.token_path and potential_end_token.token_type in JsonTokenType.end_tokens():
+                    break
+
+    async def skip_until_property_name(self, name: str):
+        while True:
+            token = await self.read_token_of_type(JsonTokenType.MAP_KEY)
+            if token.token_value == name:
+                return token
+
+            await self.skip_children(token)
+
+    async def skip_until_any_property_name(self, *names: str):
+        while True:
+            token = await self.read_token_of_type(JsonTokenType.MAP_KEY)
+            if token.token_value in names:
+                return token
+
+            await self.skip_children(token)
+
+    async def skip_until_property_name_or_end_object(self, *names: str) -> JsonToken:
+        async for token in self:
+            if token.token_type == JsonTokenType.END_MAP:
+                return token
+
+            if token.token_type == JsonTokenType.MAP_KEY:
+                if token.token_value in names:
+                    return token
+
+                await self.skip_children(token)
+                continue
+
+            raise Exception(f"Unexpected token {token}")
+
+    async def skip_until_token_with_paths(self, *tokens: (JsonTokenType, str)):
+        async for token in self:
+            if any((token.token_type == t_type and token.token_path == t_path) for (t_type, t_path) in tokens):
+                return token
+            await self.skip_children(token)
+
+
+class ProgressiveDataSetEnumerator:
```

Comment:
Amit Ofir brought to my knowledge that streaming isn't progressive - to have a progressive query you need to specify it in the ClientRequestProperties  - but this isn't the case here, so I think we better not use the term "progressive" as this is confusing anyway. Sorry if I were misleading you..

Reply:
I removed the progressive wording, supporting it seems nice but I think it's Out Of Scope for this PR

@yogilad what do you think?